bazel/utils/cc_dynamic_wrapper.bzl

```diff
@@ -0,0 +1,172 @@
+"""Rules to generate wrappers for dynmaic binaries."""
+
+load("@bazel_skylib//lib:paths.bzl", "paths")
+load("@bazel_skylib//lib:shell.bzl", "shell")
+load("//bazel/utils:messaging.bzl", "fileowner")
+load("//bazel/utils:messaging.bzl", "package")
+
+def _cc_dynamic_wrapper_impl(ctx):
+    template = """#!/bin/bash
+export LD_LIBRARY_PATH={ldpaths}:$LD_LIBRARY_PATH
+DEBUG=${{DEBUG:-{debug}}}
+test -z "$DEBUG" || {{
+  echo "===== dynamic wrapper {target} - files availale ======="
+  find -L .
+  echo "===== running {run} $@ ======="
+  echo "LD_LIBRARY_PATH=$LD_LIBRARY_PATH"
+}}
+exec {run} "$@"
+"""
+    run = ""
+    targets = ctx.attr.deps
+    if ctx.executable.bin:
+        targets = [ctx.attr.bin] + targets
+        run = shell.quote(ctx.executable.bin.short_path)
+    if ctx.attr.run:
+        run = ctx.expand_location(ctx.attr.run, targets = targets)
+
+    dirs = {}
+    runfiles = ctx.runfiles()
+    for dep in targets:
+        di = dep[DefaultInfo]
+
+        files = dep.files.to_list()
+        runfiles = runfiles.merge(ctx.runfiles(files = files))
+        if di.files_to_run and di.files_to_run.executable:
+            runfiles = runfiles.merge(ctx.runfiles(files = [di.files_to_run.executable]))
+
+        deprf = di.default_runfiles
+        if deprf and deprf.files:
+            runfiles = runfiles.merge(deprf)
+            files.extend(deprf.files.to_list())
+
+        for f in files:
+            if not f.path.endswith(".so") and not ".so." in f.path:
```

Comment:
Would it not be better to just grab `out_shared_libs`?

Reply:
Also, FWIW, for some of our internal usage, we have targets that export `.so.X` versioned libs too (which is wrong and I never understood why we needed it, but we still have them)

Reply:
I meant to say, can you not grab out_lib_dir and use it to set ldpaths in L#57

Reply:
re `out_shared_libs` is only a thing of `rules_foreign_cc`, `configure_make` and friends unfortunately, and the original attributes of a rule definition are not available to downstream rules. For example, `configure_make` would interpret its rule specific attributes and create a bunch of outputs, and annotate them with `providers`. Here, I can access the outputs and providers only - the original attributes are rule private. The proper way would be to look at the `CcInfo` provider, that `annotates all C/C++ compiled things with the linker properties`, but... that provider is only available / fully filled for .c/.c++ built through bazel, not through Make/Makefile/... (or at least, it does not look like rules_foreign_cc provides all that is needed).

In experiments, it seemed like going through the set of files brought in via dependencies or runfiles was the simplest / most reliable.

That said:
1. I've added a comment in the .bzl file explaining this.
2. `.so.X` files are captured correctly (see second half of the if)
3. Will do a couple more experiments before merging, if you have rules in mind I could test this with, that'd be awesome.

Reply:
I see, makes sense!